index.js

```diff
@@ -1,130 +1,158 @@
-  return new Promise((resolve) => {
-    const { worker } = require(outputCompiledFilename)[runModuleName];
-    let app;
+    // ensure project folder is adequate
+    try {
+      const projectDirStats = await stat(projectDir);
+      if (!projectDirStats.isDirectory()) {
+        throw new Error();
+      }
+    } catch (err) {
+      throw new Error(`Provided --project-dir '${rawProjectDir}' is not a directory`);
+    }
+    if (userModulePath.indexOf(`${projectDir}/`) !== 0) {
+      throw new Error(`File \`${userModulePath}\` must be located within --project-dir \`${rawProjectDir}\``);
+    }
 
-    if (needArgs) {
-      const args = process.argv.slice(3);
-      app = worker(args);
-    } else {
-      app = worker();
+    // read user module and determine what template to use
+    let userModuleContents;
+    try {
+      userModuleContents = await readFile(userModulePath, 'utf-8');
+    } catch (err) {
+      throw new Error(`File '${userModulePath}' could not be read`);
+    }
+    const argsRegex = /^output *\w+ *=/;
+    const needArgs = userModuleContents
+      .split('\n')
+      .some(line => argsRegex.test(line.trim()));
+
+    // read main module template
+    await sh.cd(projectDir);
+    let template;
+    const chosenTemplatePath = needArgs ? templateWithArgsPath : templatePath;
+    try {
+      template = await readFile(chosenTemplatePath, 'utf-8');
+    } catch (err) {
+      throw new Error(`Elm file '${chosenTemplatePath}' does not exist`);
     }
 
-    app.ports.sendOutput.subscribe((output) => {
-      console.log(output);
-      resolve();
+    // create main module file from template
+    const mainModuleContents = template
+      .replace('{mainModule}', mainModule)
+      .replace('{userModule}', userModule)
+      .replace('{output}', outputVarName);
+    await writeFile(mainModuleFilename, mainModuleContents, 'utf-8');
+
+    // compile main module
+    const contents = await compileToString([mainModuleFilename], { yes: true });
+    await writeFile(outputCompiledFilename, contents);
+
+    // run compiled elm file
+    await new Promise((resolve) => {
+      const { worker } = require(outputCompiledFilename)[mainModule];
+      const app = needArgs ? worker(argsToOutput) : worker();
+      app.ports.sendOutput.subscribe((output) => {
+        console.log(output);
+        resolve();
+      });
     });
-  });
-}
-
-function cleanup() {
-  return Promise.all([
-    unlink(outputElmFilename).catch(() => {}),
-    unlink(outputCompiledFilename).catch(() => {}),
-  ]);
-}
-
-function handleError(err) {
-  let message;
-
-  if (typeof err === 'object' && 'message' in err) {
-    if (/does not expose `output`/.test(err.message)) {
-      message = `Elm module \`${moduleName}\` does not define a String \`output\`.`;
+  } catch (err) {
+    // handle error
+    let message;
+    if (typeof err === 'object' && 'message' in err) {
+      if (err.message.indexOf(`does not expose \`${outputVarName}\``) !== -1) {
```

Comment:
We should probably create our own error types to make inspection cleaner here and also below.

Reply:
Maybe, but I'm not sure if this is important at the moment. Both to the "special" errors that we handle in a peculiar way are easy to understand ATM. If we add our own error types, the code may look a bit more organised, but will also introduce new abstractions that a reader would have to understand. The simpler the better IMO, we don't want to overengineer.